speechbrain/utils/distributed_metrics.py

```diff
@@ -0,0 +1,159 @@
+"""SpeechBrain utility functions for distributed metrics.
+
+This file is heavily inspired by the `huggingface accelerate` library, which is licensed under
+the Apache License 2.0. The original code can be found at: https://github.com/huggingface/accelerate/tree/main
+
+The code has been adapted to work with SpeechBrain and only for CPU and GPU training.
+
+Authors:
+ * Adel Moumen 2024
+"""
+
+from typing import Any
+
+import torch
+from packaging import version
+
+from speechbrain.utils.distributed import DistributedState
+from speechbrain.utils.distributed_utils import recursively_apply
+
+
+def _gpu_gather_object(object: Any):
+    """Gathers an object from all processes and returns a list containing the object from each process."""
+    output_objects = [None for _ in range(torch.distributed.get_world_size())]
+    torch.distributed.all_gather_object(output_objects, object)
+    # all_gather_object returns a list of lists, so we need to flatten it
+    return [x for y in output_objects for x in y]
+
+
+def _gpu_gather(tensor):
+    """Gathers a tensor from all processes and returns a tensor containing the gathered tensors."""
+    state = DistributedState()
+    if version.parse(torch.__version__) >= version.parse("1.13"):
+        gather_op = torch.distributed.all_gather_into_tensor
+    else:
+        gather_op = torch.distributed._all_gather_base
+
+    def _gpu_gather_one(tensor):
+        if tensor.ndim == 0:
+            tensor = tensor.clone()[None]
+
+        # Can only gather contiguous tensors
+        if not tensor.is_contiguous():
+            tensor = tensor.contiguous()
+
+        if (
+            state.distributed_backend is not None
+            and state.distributed_backend != "gloo"
```

Comment:
Again, couldn't we use https://pytorch.org/docs/stable/distributed.html#torch.distributed.get_backend here rather than state?

Reply:
I think you are right on that! Maybe i can wipe up the singleton. Wdyt ? (as I said in a previosu comment, I thought the singleton to be particularly useful in the future if we want to make our codebase compatible with other features like FSDP but Im fine to revert it)

Reply:
I did the changes. Thanks for your comment. :)